alf/algorithms/diayn_algorithm.py

```diff
@@ -0,0 +1,135 @@
+
+    See Eysenbach et al "Diversity is All You Need: Learning Diverse Skills without a Reward Function"
+    """
+
+    def __init__(self,
+                 num_of_skills,
+                 feature_spec,
+                 hidden_size=256,
+                 encoding_net: Network = None,
+                 discriminator_net: Network = None,
+                 name="DIAYNAlgorithm"):
+        """Create a DIAYNAlgorithm.
+
+        Args:
+            num_of_skills: number of skills
+            hidden_size (int|tuple): size of hidden layer(s)
+            encoding_net (Network): network for encoding observation into a
+                latent feature specified by feature_spec. Its input is same as
+                the input of this algorithm.
+            discriminator_net (Network): network for predicting the skill labels based on the observation.
+        """
+        super().__init__(train_state_spec=feature_spec, name=name)
+
+        flat_feature_spec = tf.nest.flatten(feature_spec)
+        assert len(flat_feature_spec
+                   ) == 1, "DIAYNAlgorithm doesn't support nested feature_spec"
+
+        self._num_skills = num_of_skills
+        self._skill_spec = tf.TensorSpec(
+            shape=(self._num_skills, ), dtype=tf.int32)
+
+        self._encoding_net = encoding_net
+
+        if isinstance(hidden_size, int):
+            hidden_size = (hidden_size, )
+
+        if discriminator_net is None:
+            discriminator_net = EncodingNetwork(
+                name="discriminator_net",
+                input_tensor_spec=feature_spec,
+                fc_layer_params=hidden_size,
+                last_layer_size=self._num_skills,
+                last_kernel_initializer=tf.initializers.Zeros())
+
+        self._discriminator_net = discriminator_net
+
+    def _encode_skill(self, skill):
+        if tensor_spec.is_discrete(self._skill_spec):
+            return tf.one_hot(indices=skill, depth=self._num_skills)
+        else:
+            return skill
+
+    def train_step(self, inputs, state, calc_intrinsic_reward=True):
+        """
+        Args:
+            inputs (tuple):  skill-augmened observation and previous action
+            state (Tensor): state for DIAYN (previous observation)
+            calc_intrinsic_reward (bool): if False, only return the losses
+        Returns:
+            TrainStep:
+                outputs: empty tuple ()
+                state: observation
+                info (DIAYNInfo):
+        """
+        observations_aug, _ = inputs
+        observation, skill = observations_aug
+
+        if self._encoding_net is not None:
+            feature, _ = self._encoding_net(observation)
+
+        prev_feature = state
+        skill_index = tf.cast(skill, tf.int32)
+        skill = self._encode_skill(skill_index)
+
+        skill_pred, _ = self._discriminator_net(inputs=feature)
+
+        if tensor_spec.is_discrete(self._skill_spec):
```

Comment:
self._skill_spec is always discrete at line 61

Reply:
Unused branch has been removed.